torch/distributed/_shard/sharding_spec/api.py

```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
             memory_format=torch.contiguous_format,
             pin_memory=tensor.is_pinned()
         )
+        current_rank = dist.get_rank(process_group)
         tensor_meta = self.build_metadata(tensor.size(), tensor_properties)
         local_shards = []
+        local_tensor = None
+        local_metadata = None
+        tensors_to_scatter = [None] * dist.get_world_size(process_group)
 
-        current_rank = dist.get_rank(process_group)
-        # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
-        dist.broadcast(tensor, src=src_rank, group=process_group)
+        sharding_dim_size = tensor.size()[self.dim]  # type: ignore[index]
+        chunks = len(self.placements)
+        split_size = get_split_size(sharding_dim_size, chunks)
+        scatter_shape = list(tensor.size())
+        scatter_shape[self.dim] = split_size  # type: ignore[index]
 
         for shard_meta in tensor_meta.shards_metadata:
             rank, device = _parse_and_validate_remote_device(process_group, shard_meta.placement)
-            if rank == current_rank:
+            if current_rank == src_rank:
                 # Reshape to get shard for this rank and we don't want autograd
                 # recording here for the narrow op and 'local_shard' should be a
                 # leaf variable in the autograd graph.
-                local_tensor = narrow_tensor(tensor, shard_meta).clone().detach().contiguous()
-
-                # Sync requires_grad to local_shard.
-                local_tensor.requires_grad = tensor.requires_grad
-                local_shards.append(
-                    Shard(
-                        tensor=local_tensor,
-                        metadata=shard_meta
-                    )
-                )
+                narrowed_tensor = narrow_tensor(tensor, shard_meta)
+                if shard_meta.shard_sizes[self.dim] < split_size:  # type: ignore[index]
+                    # for the last shard that might be smaller to other shards
+                    # resize the narrowed tensor to the same size and use it for
+                    # the scatter collective as dist.scatter requires same size
+                    # inputs on every rank
+                    tensor_to_scatter = narrowed_tensor.detach().clone().resize_(scatter_shape)
```

Comment:
I tested it locally, this is essentially same as padding, random numbers are just added into the tensor for a bigger shape.. From the wiki of `resize_`, it says "Existing elements are preserved but any new memory is uninitialized.", I am not sure if there could be any side effect. If not, I think it's good to go.

Reply:
Yeah I also looked into the `gather_object`  and `all_gather_object` in the c10d API, seems like they are all using `resize_` ([link](https://github.com/pytorch/pytorch/blob/b369b89f235f54bc9de85d768fb62ac4579681dc/torch/distributed/distributed_c10d.py#L1761)) So I think we are good here.

I am even thinking maybe we should switch the `pad` and `unpad` to also use `resize_` to consolidate our approaches on the non-uniform inputs.

Reply:
I see. Is there difference in padding and resize? I mean which one is more efficient?